{% extends 'base.html' %}

{% block static %}
    <link href="{{ url_for('static', filename='select2/css/select2.min.css') }}" rel="stylesheet" />
    {% include 'includes/loading_bar.dt.j2' %}

    <link href="{{ url_for('static', filename='seal/analysis/sample.css') }}" rel="stylesheet" type="text/css">
{% endblock %}

{% block content %}
    {% include 'includes/sidebar-sample.j2' %}
    <div class="w3-main {% if sample.status == 5 %}w3-disabled{% endif %}">
        <div class="w3-container">
            <div class="w3-center">
                <h1>{{ sample.samplename}}</h1>
                {% include 'includes/button-status-sample.j2' %}
            </div>
            <hr>
        </div>
        <div class="w3-padding w3-small">
            <div class="seal-form-report w3-row w3-padding {% if sample.status == 4 %}w3-disabled {% endif %}">
                <div class="w3-col w3-padding w3-text-flat-green-sea" style="width:200px"><i class="w3-large fa fa-filter"></i> <b>Choose a filter</b></div>
                <select id="selectFilter" class="w3-select w3-border w3-col m3 l2 s6" onchange="changeFilter(value, '{{ sample.id }}');">
                </select>

                <div class="w3-col w3-padding w3-text-flat-green-sea" style="width:200px"><i class="w3-large fas fa-th-list"></i> <b>Choose a panel</b></div>
                <select id="selectBed" class="w3-select w3-border w3-col m3 l2 s6" onchange="applied_panel(value, '{{ sample.id }}');">
                </select>
            </div>
            <div class="w3-padding">
                <table id="variants" class="w3-table display" style="width:100%">
                    <thead>
                        <tr>
                            <th rowspan="2" class='min-size-75'><span class="seal-display-middle">Gene</span></th>
                            <th rowspan="2" class='min-size-75'><span class="seal-display-middle"></span></th>
                            <th colspan="4" class="w3-border-right">Nomenclature</th>
                            <th colspan="4" class="w3-border-right">Variant Quality</th>
                            <th colspan="4" class="w3-border-right">Population Frequency</th>
                            <th colspan="2" class="w3-border-right">Clinical Informations</th>
                            <th colspan="2" class="w3-border-right">Localisation</th>
                            <th colspan="3" class="w3-border-right">Predictions</th>
                            <!-- <th colspan="3">Contact</th> -->
                            <th  rowspan="2" class='min-size-2x5'><span class="seal-display-middle">R</span></th>
                            <th  rowspan="2" class='min-size-75'><span class="seal-display-middle">Class</span></th>
                            <th  rowspan="2" class='min-size-25'></th>
                        </tr>
                        <tr>
                            <th class='min-size-75'>HGVSg</th>
                            <th class='min-size-75'>HGVSc</th>
                            <th class='min-size-75'>HGVSp</th>
                            <th class='min-size-50'>E|I</th>
                            <th class='min-size-75'>Filter</th>
                            <th class='min-size-50'>AF</th>
                            <th class='min-size-50'>DP</th>
                            <th class='min-size-50'>AD</th>
                            <th class='min-size-75'>GnomAD</th>
                            <th class='min-size-75'>SEAL %</th>
                            <th class='min-size-75'>SEAL #</th>
                            <th class='min-size-75'>In Family</th>
                            <th class='min-size-75'>ClinSig</th>
                            <th class='min-size-75'>Transmission</th>
                            <th class='min-size-75'>Impact</th>
                            <th class='min-size-75'>Consequence</th>
                            <th class='min-size-50'>MaxEnt</th>
                            <th class='min-size-50'>SpliceAI</th>
                            <th class='min-size-100'>Missense</th>
                        </tr>
                    </thead>
                    <tfoot>
                        <tr>
                            <th>
                                <span onclick="openHelpModal(title='Gene', content='The gene symbol (e.g. HGNC) (where available) to the output.', footer='For more information see : <a  href=\'https://www.genenames.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>HGNC</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Gene
                            </th>
                            <th></th>
                            <th>
                                <span onclick="openHelpModal(title='HGVSg', content='The HGVS genomic sequence name', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                HGVSg
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='HGVSc', content='The HGVS coding sequence name.<br /><br />Canonical transcript is defined by Ensembl algorithm following :<br /><ol><li>Longest CCDS translation with no stop codons.</li><li>If no (1), choose the longest Ensembl/Havana merged translation with no stop codons.</li><li>If no (2), choose the longest translation with no stop codons.</li><li>If no translation, choose the longest non-protein-coding transcript.</li></ol><br /><ul><li>the transcript is the canonical transcript for the gene<ul><li><i class=\'fas fa-star\'></i> (<i>canonical</i>)</li><li><i class=\'far fa-star\'></i> (<i>non canonical</i>)</li></ul></li><li>the transcript is one of your selected transcripts<ul><li><i class=\'w3-text-flat-alizarin fas fa-tint\'></i> (<i>selected</i>)</li><li><i class=\'w3-text-flat-peter-river fas fa-tint\'></i> (<i>not selected</i>)</li></ul></li></ul>', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>, <a  href=\'http://www.ensembl.org/info/genome/genebuild/canonical.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a> & <a  href=\'{{ url_for('transcripts') }}\' target=\'_blank\' style=\'text-decoration:underline\'><i>Transcripts in SEAL</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                HGVSc
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='HGVSp', content='The HGVS protein sequence name', footer='For more information see : <a  href=\'https://varnomen.hgvs.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>varnomen</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                HGVSp
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Exon/Intron', content='The number of the exon or intron wich have the variant.', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                E|I
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Filter', content='Filters applied to the variant.', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Filter
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Allelic frequencie', content='Frequence of the observed allele for the sample (ie: Allelic depth/Depth)<br />Genotypes :<ul><li><i class=\'far fa-dot-circle\'></i> Mosaicism (< 0.25)</li><li><i class=\'fas fa-adjust\'></i> Heterozygous ref/alt (< 0.75)</li><li><i class=\'fas fa-circle\'></i> Homozygous alt/alt (>= 0.75)</li><li><i class=\'far fa-question-circle\'></i> Unknown status</li></ul>', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                AF
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Depth', content='Depth in DNA sequencing is the number of unique reads that include a given nucleotide in the reconstructed sequence.', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                DP
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Allelic depth', content='Depth of the observed allele.', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                AD
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='GnomAD', content='The Genome Aggregation Database (gnomAD) is a resource developed by an international coalition of investigators, with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects, and making summary data available for the wider scientific community.<br /> The dataset used is from release 2.1 genomes (GRCh37/hg19) spans 15,708 whole-genome sequences from unrelated individuals sequenced as part of various disease-specific and population genetic studies.', footer='For more information see : <a  href=\'https://gnomad.broadinstitute.org\' target=\'_blank\' style=\'text-decoration:underline\'><i>GnomAD website</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                <i class="fas fa-question-circle w3-tiny"></i>
                            </span>
                                GnomAD
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='SEAL (freq)', content='The frequency with which this variant is observed in SEAL among all other samples.', footer='Warning : this value exclude this sample')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                SEAL %
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='SEAL (count)', content='The frequency with which this variant is observed in SEAL among all other samples.', footer='Warning : this value exclude this sample')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                SEAL #
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='In Family', content='Count variant presence in the case family and list all samples in the faimily with this variant.', footer='')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                In Family
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='ClinSig', content='ClinVar clinical significance of the dbSNP variant.', footer='For more information see : <a  href=\'https://www.ncbi.nlm.nih.gov/clinvar/docs/clinsig/\' target=\'_blank\' style=\'text-decoration:underline\'><i>ClinSig documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                ClinSig
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Transmission', content='This column represent the transmission mode following OMIM data.<br /><ul><li><b>AD</b>: Autosomal dominant</li><li><b>AR</b>: Autosomal recessive</li><li><b>DD</b>: Digenic dominant</li><li><b>DR</b>: Digenic recessive</li><li><b>IC</b>: Isolated cases</li><li><b>Mito</b>: Mitochondrial</li><li><b>MF</b>: Multifactorial</li><li><b>PAD</b>: Pseudoautosomal dominant</li><li><b>PAR</b>: Pseudoautosomal recessive</li><li><b>SMos</b>: Somatic mosaicism</li><li><b>SMut</b>: Somatic mutation</li><li><b>XL</b>: X-linked</li><li><b>XLD</b>: X-linked dominant</li><li><b>XLR</b>: X-linked recessive</li><li><b>YL</b>: Y-linked</li></ul><br /><i class=\'fas fa-question-circle w3-tiny\'></i> A question mark (?) indicates that the relationship between the phenotype and gene is provisional', footer='For more information see : <a  href=\'https://www.omim.org/\' target=\'_blank\' style=\'text-decoration:underline\'><i>OMIM</i></a>' )" class="w3-opacity-max w3-hover-opacity">
                                <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Transmission
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Impact', content='IMPACT is calculated by VEP, based on consequences and compatibility with snpEff. <div class=\'w3-flash w3-warning\'><p>Impact categories must be used with care, they were created only to help and simplify the filtering process. Obviously, there is no way to predict whether a HIGH impact or a LOW impact variant is the one producing a phenotype of interest.</p></div>', footer='For more information see : <a  href=\'https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a> & <a  href=\'https://pcingola.github.io/SnpEff/se_inputoutput/#impact-prediction\' target=\'_blank\' style=\'text-decoration:underline\'><i>snpEff documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Impact
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Consequence', content='Consequences are calculated by VEP following SO terms.', footer='For more information see : <a  href=\'https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP documentation</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Consequence
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='MaxEntScan', content='Algorithms for derivation and scoring of constrained-marginal maximum entropy distributions, with applications to 5\' and 3\' splice sites. <br />We suggest this cutoff: <br /><ul><li class=\'w3-text-flat-pomegranate\'>|percent variation| &ge; 15\%</li></ul>', footer='For more information see : <a  href=\'https://github.com/guillermomarco/VEP_plugins/blob/master/MaxEntScan.pm\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP MES plugin</i></a> & <a  href=\'https://github.com/esebesty/maxentscan\' target=\'_blank\' style=\'text-decoration:underline\'><i>MaxEntScan</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                MaxEnt
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='SpliceAI DeltaScore', content='Delta score (DS) of a variant, defined as the maximum of ranges from 0 to 1 and can be interpreted as the probability of the variant being splice-altering. The author-suggested cutoffs are: <br /><ul><li class=\'w3-text-flat-sun-flower\'>0.2 (high recall)</li><li class=\'w3-text-flat-orange\'>0.5 (recommended)</li><li class=\'w3-text-flat-pomegranate\'>0.8 (high precision)</li></ul>', footer='For more information see : <a  href=\'https://github.com/Ensembl/VEP_plugins/blob/release/104/SpliceAI.pm\' target=\'_blank\' style=\'text-decoration:underline\'><i>VEP plugins spliceAI</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                SpliceAI
                            </th>
                            <th>
                                <span onclick="openHelpModal(title='Missense', content='This score represents the mean rank scores of the top 10 missense predictions tools from dbNSFP v4.<br /><ul><li>CADD</li><li>VEST4</li><li>MetaSVM</li><li>MetaLR</li><li>Eigen</li><li>Eigen-PC</li><li>REVEL</li><li>BayesDel_addAF</li><li>BayesDel_noAF</li><li>ClinPred</li></ul><br />The threshold are : <ul><li><span class=\'w3-tag w3-flat-pomegranate\'>0.95</span> <i>top 5% deleteriousness of all missense variant</i></li><li><span class=\'w3-text-flat-pomegranate\'>0.9</span> <i>top 10% deleteriousness of all missense variant</i></li><li><span class=\'w3-text-flat-orange\'>0.75</span> <i>top 25% deleteriousness of all missense variant</i></li><li><span class=\'w3-text-flat-sun-flower\'>0.5</span> <i>top 50% deleteriousness of all missense variant</i></li></ul>', footer='For more information see : <a  href=\'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7709417/\' target=\'_blank\' style=\'text-decoration:underline\'><i>dbNSFP v4</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Missense
                            </th>
                            <th>
                                R
                            </th>
                            <th style="padding:10px 10px 6px 10px!important">
                                <span onclick="openHelpModal(title='Class', content='You can choose the classification:<ul><li class=\'w3-text-flat-green-sea\'>Benign</li><li class=\'w3-text-flat-nephritis\'>Likely Benign</li><li class=\'w3-text-flat-belize-hole\'>VUS</li><li class=\'w3-text-flat-pumpkin\'>Likely Pathogenic</li><li class=\'w3-text-flat-pomegranate\'>Pathogenic</li></ul>', footer='For more information see : <a  href=\'https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/\' target=\'_blank\' style=\'text-decoration:underline\'><i>ACMG guidelines</i></a>')" class="w3-opacity-max w3-hover-opacity">
                                    <i class="fas fa-question-circle w3-tiny"></i>
                                </span>
                                Class
                            </th>
                            <th></th>
                        </tr>
                    </tfoot>
                </table>
            </div>
        </div>
        <div class="w3-container w3-center">
            <hr>
        </div>

        <div class="w3-container w3-padding">
            <div id="comments_sample" class=" w3-container w3-border-right">
                <h2><i class="w3-text-flat-green-sea w3-xxlarge fas fa-comment"></i> Comments (<span class="commentsCountSample">{{ sample.comments | length }}</span>)</h2>

                <div id="commentsTableSample" class="w3-small">
                    <table id="tableCommentsSample" class="display" style="width:100%">
                        <thead>
                            <tr>
                                <th class="no-sort" style="width:20%">User</th>
                                <th class="no-sort" style="width:60%">Comment</th>
                                <th style="width:20%">Date</th>
                            </tr>
                        </thead>
                    </table>
                </div>

                {{ form.hidden_tag() }}
                <div class="w3-row w3-section">
                    <div class="w3-rest">
                        {% if form.comment.errors %}
                            {{ form.comment(class="w3-input w3-border w3-border-red") }}
                            <div class="w3-small w3-text-flat-alizarin">
                                {% for error in form.comment.errors %}
                                    <i class="fas fa-exclamation-circle"></i> {{ error }}
                                {% endfor %}
                            </div>
                        {% else %}
                            {{ form.comment(id="comment_sample", class="w3-input w3-border") }}
                        {% endif %}
                    </div>
                </div>

                <div class="w3-section w3-center">
                    {{ form.submit(id="submitCommentSample", class="w3-button w3-flat-green-sea w3-hover-flat-turquoise", onclick="send_comment();") }}
                </div>
            </div>
        </div>
        <div class="w3-container w3-center">
            <hr>
        </div>
    </div>

    <!-- MODALS -->
    <div id="help-modal" class="w3-modal">
        <div class="w3-modal-content w3-card-4">
            <header class="w3-container w3-flat-green-sea">
                <span onclick="$('#help-modal').toggle()"
                    class="w3-button w3-display-topright">&times;</span>
                <h5 id="help-modal-title"></h5>
            </header>
            <div class="w3-container">
                <p id="help-modal-content"></p>
            </div>
            <footer class="w3-container w3-flat-turquoise">
                <p id="help-modal-footer"></p>
            </footer>
        </div>
    </div>

    <div id="filter-modal" class="w3-modal">
        <div class="w3-modal-content w3-card-4">
            <header class="w3-container w3-flat-green-sea">
                <span onclick="$('#filter-modal').toggle()"
                    class="w3-button w3-display-topright">&times;</span>
                <h5 id="filter-modal-title">Save Filter</h5>
            </header>
            <div class="w3-container">
                {{ saveFilterForm.hidden_tag() }}
                <div class="w3-row w3-section">
                    <div class="w3-col w3-display-container w3-text-flat-green-sea w3-margin-right" style="width:50px"><i class="w3-xxlarge fas fa-tag"></i></div>
                    <div class="w3-rest">
                        {% if saveFilterForm.filterName.errors %}
                            {{ saveFilterForm.filterName(class="w3-input w3-border w3-border-red") }}
                            <div class="w3-small w3-text-flat-alizarin">
                                {% for error in saveFilterForm.filterName.errors %}
                                    <i class="fas fa-exclamation-circle"></i> {{ error }}
                                {% endfor %}
                            </div>
                        {% else %}
                            {{ saveFilterForm.filterName(class="w3-input w3-border") }}
                        {% endif %}
                    </div>
                </div>

                <div class="w3-row w3-section w3-display-container">
                    <div class="w3-middle w3-col w3-display-container w3-text-flat-green-sea w3-margin-right" style="width:50px"><i class="w3-xxlarge fas fa-users"></i></div>
                    <div class="w3-rest" id="create-filter">
                        {{ saveFilterForm.teams(class="js-example-basic-multiple", multiple="multiple", style="width:100%") }}
                    </div>
                </div>
                <div class="w3-row w3-section">
                    <div class="w3-col w3-display-container w3-text-flat-green-sea w3-margin-right" style="width:50px"><i class="w3-xxlarge fas fa-filter"></i></div>
                    <div class="w3-rest">
                        {% if saveFilterForm.filterText.errors %}
                            {{ saveFilterForm.filterText(class="w3-input w3-border w3-border-red") }}
                            <div class="w3-small w3-text-flat-alizarin">
                                {% for error in saveFilterForm.filterText.errors %}
                                    <i class="fas fa-exclamation-circle"></i> {{ error }}
                                {% endfor %}
                            </div>
                        {% else %}
                            {{ saveFilterForm.filterText(class="w3-input w3-border", disabled=True) }}
                        {% endif %}
                    </div>
                </div>

                <div class="w3-section w3-center">
                    {{ saveFilterForm.submit(id="validateSaveFilter", class="w3-button w3-flat-green-sea w3-hover-flat-turquoise", onclick="saveFilter();") }}
                </div>
            </div>
            <footer class="w3-container w3-light-grey w3-padding">
                <button class="w3-button w3-right w3-white w3-border"
                onclick="$('#filter-modal').toggle()">Close</button>
            </footer>
        </div>
    </div>

    <div id="modal-details-variant" class="w3-modal">
        <div class="w3-modal-content w3-card-4 w3-animate-zoom">
            <header class="w3-container w3-flat-green-sea">
                <span onclick="$('#modal-details-variant').toggle()"
                class="w3-button w3-xlarge w3-display-topright">&times;</span>
                <h2>Variant : <span id="variant-id"><span></h2>
            </header>

            <div class="w3-bar w3-border-bottom">
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'allTranscripts')">Transcripts</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'missense')">Missense</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'conservation')">Conservation</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'population')">Population</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'splicing')">Splicing</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'links')">Links</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'inseal')">In Seal (<span class="inSealCount"></span>)</button>
                <button class="tablink w3-bar-item w3-button" onclick="openDetails(event, 'comments')">Comments (<span class="commentsCountVar"></span>)</button>
            </div>

            <div id="allTranscripts" class="w3-container detail">
                <h3>Transcripts</h3>
                <div id="allTranscriptsTable"></div>
            </div>

            <div id="missense" class="w3-container detail">
                <h3>Missense</h3>
                <div class="w3-row">
                    <div id='radarMissense' class="w3-col l6 m12 s12"><!-- Plotly chart will be drawn inside this DIV --></div>
                    <div class="w3-col l6 m12 s12">
                        <div id="tableScoresPredictions"></div>
                    </div>
                </div>
            </div>

            <div id="conservation" class="w3-container detail">
                <h3>Conservation</h3>
                <div class="w3-row">
                    <div id='radarConservation' class="w3-col l6 m12 s12"><!-- Plotly chart will be drawn inside this DIV --></div>
                    <div class="w3-col l6 m12 s12">
                        <div id="tableScoresConservation"></div>
                    </div>
                </div>
            </div>

            <div id="population" class="w3-container detail">
                <h3>Population</h3>
                <div class="w3-row">
                    <div id='radarPopulation' class="w3-col l6 m12 s12"><!-- Plotly chart will be drawn inside this DIV --></div>
                    <div class="w3-col l6 m12 s12">
                        <div id="tableScoresPopulation"></div>
                    </div>
                </div>
            </div>

            <div id="splicing" class="w3-container detail">
                <h3>Splicing</h1>
                <div class="w3-row">
                    <div id='chartSplicing' class="w3-col l3 m6 s6"><!-- Plotly chart will be drawn inside this DIV --></div>
                </div>
            </div>

            <div id="links" class="w3-container detail">
                <h3>Links</h3>
                <div id="links_exterior" style="column-count: 3;"></div>
                <div class="w3-padding w3-display-bottommiddle w3-tiny">
                    <span><i class="w3-text-flat-asbestos fas fa-flag"></i> dbSNP</span>
                    <span><i class="w3-text-flat-amethyst fas fa-disease"></i> Cosmic</span>
                    <span><i class="w3-text-flat-belize-hole fas fa-user-md"></i> ClinVar</span>
                    <span><i class="w3-text-flat-pomegranate fas fa-microscope"></i> UniProt</span>
                    <span><i class="w3-text-flat-wet-asphalt fas fa-book"></i> PubMed</span>
                </div>
            </div>

            <div id="inseal" class="w3-container detail">
                <h3>In Seal (<span class="inSealCount"></span>)</h3>
                <div id="inSEALTable"></div>
            </div>

            <div id="comments" class="w3-container detail">
                <h3>Comments (<span class="commentsCountVar"></span>)</h3>

                {{ form.hidden_tag() }}
                <div class="w3-row w3-section">
                    <div class="w3-col w3-display-container w3-text-flat-green-sea w3-margin-right" style="width:50px"><i class="w3-xxlarge fas fa-comment"></i></div>
                    <div class="w3-rest">
                        {% if form.comment.errors %}
                            {{ form.comment(class="w3-input w3-border w3-border-red") }}
                            <div class="w3-small w3-text-flat-alizarin">
                                {% for error in form.comment.errors %}
                                    <i class="fas fa-exclamation-circle"></i> {{ error }}
                                {% endfor %}
                            </div>
                        {% else %}
                            {{ form.comment(id="commentVar", class="w3-input w3-border") }}
                        {% endif %}
                    </div>
                </div>

                <div class="w3-section w3-center">
                    {{ form.submit(id="submitComment", class="w3-button w3-flat-green-sea w3-hover-flat-turquoise", onclick="send_comment(type='var');") }}
                </div>
                <div id="commentsTableVar"></div>
            </div>

            <div class="w3-container w3-light-grey w3-padding">
                <button class="w3-button w3-right w3-white w3-border"
                onclick="$('#modal-details-variant').toggle()">Close</button>
            </div>
        </div>
    </div>

    <div id="MD-message-modal" class="w3-modal">
        <div class="w3-modal-content w3-card-4 w3-animate-zoom">
            <header id="MD-modal-header" class="w3-container w3-flat-green-sea">
                <span onclick="$('#MD-message-modal').toggle()"
                class="w3-button w3-xlarge w3-display-topright">&times;</span>
                <h2>MobiDetails</h2>
            </header>

            <div class="w3-container">
                <p id="message-mobidetails"></p>
            </div>

            <div class="w3-container w3-light-grey w3-padding">
                <button class="w3-button w3-right w3-white w3-border"
                onclick="$('#MD-message-modal').toggle()">Close</button>
            </div>
        </div>
    </div>

{% endblock %}


{% block script %}
    <script type="text/javascript" src="{{ url_for('static', filename='plotly-2.3.1.min.js') }}"></script>
    <script src="{{ url_for('static', filename='select2/js/select2.full.min.js') }}"></script>
    <script src="{{ url_for('static', filename='DOMPurify/purify.min.js') }}"></script>

    <script type="text/javascript">
        var sample_filter_id = "{{ sample.filter_id }}";
        var current_user_filter_id = "{{ current_user.filter_id }}";
        var sample_bed_id = "{{ sample.bed_id }}";
        var current_user_bed_id = "{{ current_user.bed_id }}";
        var sample_variants_length = "{{ sample.variants | length  }}";
        var json_variants = '/json/variants/sample/{{ sample.id }}{% if sample.bed_id %}/bed/{{ sample.bed_id }}{% endif %}';
        var sample_id = "{{ sample.id }}";
        var current_user_api_key_md = "{{ current_user.api_key_md }}";
        var current_user_transcripts = "{{ current_user.transcripts |Â safe }}";
        var sample_status = "{{ sample.status }}";
        var disabled_class = "{% if sample.status == 4 %}w3-disabled {% endif %}";
    </script>

    <script type="text/javascript" src="{{ url_for('static', filename='seal/analysis/sample.js') }}"></script>
{% endblock %}
